### PR TITLE
mgr/volumes: cleanup libcephfs handles on plugin shutdown

### PR DESCRIPTION
**Pull request by vshankar**
https://tracker.ceph.com/issues/42299

[Draft PR for now]

This changes how signals (SIGTERM, SIGINT) are handled in mgr -- handling these signals is moved to Mgr.cc source.

### CODE REVIEW

**Comment by vshankar** [comment]
> I haven't looked at this yet but wanted to ask you also undo #29169 as part of this fix: this change should make the whitelist entry unnecessary.

I deliberately didn't undo that -- wouldn't that be kind of required if ceph-mgr gets SIGKILLed?

I haven't checked if we have tests that SIGKILLs ceph-mgr. If there are none, whitelisting can be reverted.

**Comment by vshankar** [comment]
@batrick reverted commit 0060f1c -- tests do not seem to force kill ceph-mgr.

**Comment by batrick** [comment]
@vshankar revert needs signed-off-by

**Comment by vshankar** [comment]
@batrick done.

File: src/mgr/Mgr.cc [hunk 1]
```diff
@@ -219,12 +219,35 @@ std::map<std::string, std::string> Mgr::load_store()
   return loaded;
 }
 
+void Mgr::handle_signal(int signum)
+{
+  ceph_assert(signum == SIGINT || signum == SIGTERM);
+  derr << "*** Got signal " << sig_str(signum) << " ***" << dendl;
+  shutdown();
```
Comment (tchaikov) [inline src/mgr/Mgr.cc]: can we add a method to `PyModuleRegistry`, so we can also shutdown and reset `standby_modules` as well?

**Comment by batrick** [comment]
@vshankar after addressing above comments. Please also run through related RADOS suite to make sure the tests do not fail due to the unwhitelisted warning.

File: src/mgr/Mgr.cc (hunk 1 above)
Comment (vshankar) [reply src/mgr/Mgr.cc]: There's already there: https://github.com/ceph/ceph/blob/master/src/mgr/PyModuleRegistry.cc#L224

We can invoke that from `Mgr::shutdown()` and probably call that as `standby_shutdown()`?

Comment (tchaikov) [reply src/mgr/Mgr.cc]: no, we cannot. see https://tracker.ceph.com/issues/38621

Comment (vshankar) [reply src/mgr/Mgr.cc]: fair enough -- so, no other changes required to this PR, right?

Comment (tchaikov) [reply src/mgr/Mgr.cc]: i still suggest https://github.com/ceph/ceph/pull/30890#discussion_r336576088 , but i don't think it's a blocker. please feel free to ignore it, if you think it's not a must. in that case, i'll post a follow-up PR to address it.

**Comment by vshankar** [comment]
http://pulpito.ceph.com/vshankar-2019-10-22_07:53:35-rados-wip-40867-testing-basic-mira/

File: src/pybind/mgr/volumes/fs/volume.py [hunk 2]
```diff
@@ -150,19 +156,30 @@ def put_fs_handle(self, fs_name):
...
-                conn.put_fs_handle()
+                conn.put_fs_handle(notify=lambda: self.cond.notifyAll())
 
-    def _del_fs_handle(self, fs_name):
+    def _del_fs_handle(self, fs_name, wait=False):
         conn = self.connections.pop(fs_name, None)
         if conn:
-            conn.del_fs_handle()
-    def del_fs_handle(self, fs_name):
+            conn.del_fs_handle(waiter=None if not wait else lambda: self.cond.wait())
```
[the remark this answers is not in the document]
Comment (vshankar) [reply src/pybind/mgr/volumes/fs/volume.py]: done

**Comment by batrick** [comment]
@vshankar please run through fs suite with `--filter volumes` as well. If it is green, go ahead and merge.

**Comment by tchaikov** [comment]
@vshankar what did you change after you scheduled vshankar-2019-10-22_07:53:35-rados-wip-40867-testing-basic-mira ?

**Comment by vshankar** [comment]
> @vshankar what did you change after you scheduled vshankar-2019-10-22_07:53:35-rados-wip-40867-testing-basic-mira ?

rebase and @ajarr minor comment. sorry, should have mentioned during the update.

**Comment by vshankar** [comment]
http://pulpito.ceph.com/vshankar-2019-11-05_04:32:58-fs-wip-40867-testing-basic-mira/

**Comment by batrick** [comment]
See also #31398.

**Comment by vshankar** [comment]
> http://pulpito.ceph.com/vshankar-2019-11-05_04:32:58-fs-wip-40867-testing-basic-mira/

test failure is not due to the changes in this PR.

**Comment by vshankar** [comment]
> > http://pulpito.ceph.com/vshankar-2019-11-05_04:32:58-fs-wip-40867-testing-basic-mira/
> 
> test failure is not due to the changes in this PR.

It's a test case bug -- I'll send a PR for that.

**Comment by batrick** [comment]
Testing again with #31398.

**Pull request merged**